### PR TITLE
Collab with commits

### PR DESCRIPTION
**Pull request by sydneypalumbo**
Notes:
- I still need to write a solid new set of tests for this (the old ones still pass) but I wanted to get this up as soon as possible to get eyes on it
- this repo is public, just fyi
- I've sketched out how this would work in the notebook [here](https://github.com/observablehq/observablehq/compare/syd/fix-merge-edits) (which also required some codemirror upgrades) but have only verified that it should compile (with `yarn link`) and haven't done any manual or unit testing yet (top of my todo list). Also required a small upgrade to our lezer fork [here](https://github.com/observablehq/lezer/compare/main...observablehq:syd/parser-top-node).
- Skimming the above^ might give you a better sense of how I'm envisioning simplifying notebook state management with the changes here / be useful for reviewing this + its general direction (but like I said, haven't tested it by any means).

**TLDR:** Coordination between saving notebook commits and saving changes to editor states is really tricky. The goal of this PR is to alter the collab module so that it allows us to more gracefully handle this complexity seamlessly rather than hacking around it and making state management in the notebook more complicated. This is primarily done here by introducing the concept of commits: points at which the local editor state value is not yet confirmed by the authority but is effectively saved as it will override remote edits as soon as it is confirmed.

More detail:
- Previously, `collabState` contained only two properties: `version`, the version of the last saved update confirmed by the authority, and `unconfirmed`, an array of pending `LocalUpdates` awaiting confirmation by the authority. The following changes are made here:
     - `unconfirmed` renamed to `unconfirmedUpdates` for clarity
     - `unconfirmedCommits` added to track commits that have yet to be verified by the server. Commits are represented by the count of `unconfirmedUpdates` they consist of.
     - `confirmedValue` added to track the value of the document currently verified by the authority
     - `confirmedToUnconfirmed` added to track the difference between the current `confirmedValue` and the `confirmedValue` at the time of the first unconfirmed commit; allows us to map newly received remote updates over unconfirmed updates by providing a means through which to make local updates start in the same document as remote updates. (represented by a ChangeSet)
- A new `createCommit` function is exported to allow the client to mark a local commit. This adds an entry to `unconfirmedCommits` based on the current count of `unconfirmedUpdates` that are not already accounted for by existing commits.
- A new `confirmCommits` function is exported to allow the client to mark the confirmation of any number of commits by the authority. This removes entries from `unconfirmedCommits` as well as their corresponding entries in `unconfirmedUpdates`, increments the `version` according to the number cleared from the latter, and updates the `confirmedValue` by applying all newly confirmed updates to its previous value.
- The `receiveUpdates` function is modified to handle the case in which remote updates are received but there are already `unconfirmedCommits`. In this case, we still increment the `version` and use the received updates to set the `confirmedValue`, but since we know our `unconfirmedCommits` will blow away the remote changes once confirmed by the authority, we do something special to preserve them: we map the remote updates over local `unconfirmedUpdates` and then "replay" them on top of our editor state as if they were our own, effectively adding them to `unconfirmedUpdates` in a revised form after our own. Using and updating the `confirmedToUnconfirmed` ChangeSet is crucial to both (1) map remote updates we've just received and (2) be capable of receiving subsequent remote updates.
- A helper function `getConfirmedValue` is exposed to obtain the last saved value of the editor. (This will allow us generate updates when we receive remote commits.) Related, there is no concept of remote commits in these changes: remote commits should be translated into ChangeSets and received as any other remote updates would be (so we never have to deal with full string values here).
- Note that full values are not passed with commits. So one should only ever commit if the desired committed value matches the current editor value. This is a feature, not a bug, as it requires the editor state to be kept in sync with one's local model.
- In theory, I think these changes should give us flexibility to change the way we're sending edits and commits to the server later down the road, should we ever want to, because we won't simply drop useful local update info when a commit is made (as we are today).

### CODE REVIEW

File: src/collab.ts [hunk 1]
```diff
@@ -108,55 +163,228 @@ export function receiveUpdates(
...
+  // Before updating the confirmed value, generate the inverted ChangeSet.
+  const invertedChanges = changes.invert(confirmedValue)
+
+  // Update the confirmed value with remote changes.
+  confirmedValue = changes.apply(confirmedValue)
+
+  // If we're sitting on unconfirmed commits, we (1) map remote updates over
+  // unconfirmed updates, first using confirmedToUnconfirmed (2) treat these
+  // mapped remote updates as our own to prevent them from being obliterated by
+  // our commits and (3) update confirmedToUnconfirmed based on remote changes.
```
Comment (annie) [inline src/collab.ts:238]: i'm gonna try to re-explain some stuff to see how much i'm understanding correctly 😅 

> (2) treat these mapped remote updates as our own to prevent them from being obliterated by our commits

what i understand from this is that if i made a commit and then made some edits, the server handles those gracefully because they occur in a sequence and build on one another. however, if i made a commit and someone else made edits at the same time, the server could potentially blow away that other person's edits because they might be conflicting? i guess i don't understand why the server can't just map the other person's edits over my commit, but maybe that has to do with how the model works? (i should probably just look at the server code again more closely)

and moving a step forward, once we have the remote updates mapped over our local updates, we now have a sequence of updates that all build off one another, and they'll all get confirmed in order. is that right?

File: src/collab.ts [hunk 2]
```diff
@@ -108,55 +163,228 @@ export function receiveUpdates(
...
+  if (collabState.unconfirmedCommits.length) {
+    // Unlike other collab transactions, this one makes changes that should be
+    // "replayed" in our editor and thus treated as local updates, rather than
+    // ignored in the collab StateField.
+    collabAnnotation = collabReplay
+
+    // Accumulate the local changes, starting with the confirmedToUnconfirmed
+    // translation, to create a ChangeSet over which to map remote updates.
```
Comment (annie) [inline src/collab.ts:246]: << thinking out loud >>

the reason why we need to start with the `confirmedToUnconfirmed` translation is because the remote updates and our local updates are essentially on two different branches of the editor value, which diverged at the point of the first unconfirmed commit.

our local changes build off of the first unconfirmed commit, while the remote updates build off of the latest confirmed value (which may be the same as the confirmed value at the time of the first unconfirmed commit, or it may have been updated by another client (i'm very unsure if i'm understanding this part correctly!)). therefore, we basically need to rebase the remote updates onto our local updates, and to do that we need to start with a translation from the base of the remote changes (current confirmed value) to the base of our local changes (confirmed value at time of first unconfirmed commit).

is any of that right??

**Review by annie — COMMENTED** [review]
this is super cool!! i feel like i'm learning a lot of valuable stuff about multiplayer from trying to wrap my head around this 👌 

the main part that i'm still struggling to understand is the "replaying" of edits. i think this might just be because i don't fully understand why commits and edits are handled differently on the server (and what exactly the differences are!). i left a couple comments attempting to explain what i *think* is going on, so hopefully those will be helpful in showing the gaps in my understanding!

also like Bryan said, i'm down to do a call if this stuff would be easier to explain in person 👍

File: src/collab.ts (hunk 1 above)
Comment (sydneypalumbo) [reply src/collab.ts:263]: yeah this is a super good question. So when you make a commit locally, it adds an event to your local model that will alter the major version and will soon be sent to the server. But let's say that after you've made that commit, but before it's actually been confirmed by the server, we receive remote updates (what we call edits, but using updates to be consistent with the code here). We technically know how to map our locally committed updates over the remote updates, but we also know that the user committed the value of the document at some previous point, and we want to honor that so that, assuming we eventually get through to the server, that committed value will always be a part of notebook history. But in doing so, we also know that the server will update the saved document value to be our committed value. This is good for our user, but it may blow away any previously saved but uncommitted updates (like the ones we received after our commit). Instead of choosing to effectively erase those updates, we do our best to replay them on top of our edits (as you described) so they aren't lost.

In other words, the complication here comes from the fact that the model, which preserves notebook history as we know it , locks committed values in time. On one hand, this is pretty nice for users because it means they'll always be able to go back to the notebook exactly as they see it after their commit. We could, if we wanted, choose to rewrite their locally committed values when we receive remote updates and then send the revised committed values to the server. But this would mean that the notebook would not look as it did when the commit was originally locally made, so it might not be what they intended to save (and it might not even compile at all depending on how the remote updates are merged with local ones). That said, I don't think that would be crazy to do; after all, it still would be considering the updates of all users. But it also would be essentially creating artificial points in time that may not be coherent.

So we kind of try to get the best of both worlds: we still incorporate the changes of every user, but we leave previous commits unaltered, give them priority, and make the user responsible for deciding what they want to do with those remote updates in light of their own commits.

File: src/collab.ts (hunk 2 above)
Comment (sydneypalumbo) [reply src/collab.ts:271]: yes that's totally right!! going to repeat it back in different words in case that's helpful, but all sounds good to me!

Paraphrase: So at the time at which a client makes a commit (and has no previously unconfirmed commits) all of its unconfirmed updates start from the last confirmed value (i.e. they are directly applicable, in order, to the last confirmed value). The next time we receive a remote update, we know that it also starts in the last confirmed value, so we can just map the remote update over the local updates that make up the local commit. Great. Easy. But then we might receive another set of remote updates before our commit is confirmed. This time, the remote updates do _not_ originate in the same document that local updates start in, they originate in the document resulting from applying the previous remote updates to the confirmed value known at the time of the local commit. Like you said, our branches have diverged, but we want to be able to replay the remote updates over our local updates, so we need to get the two ChangeSets to a point they both understand, which is where `confirmedToUnconfirmed` comes in. By keeping track of the inversion of each remote update as it comes in, it gives us a way to start our local updates from the same document as the remote updates (i.e. the inversion itself starts at the same point as the remote updates, and the inversion ends at the point in which our local updates start, so the remote updates can be mapped over the inversion _and_ our local updates). (Because we don't actually send the updates associated with each commit when we save commits, we don't need to send the inversion, but if we did do that, we'd have to send the inversion as a ChangeSet too...)

**Comment by sydneypalumbo** [comment]
**Update**: okay so I added a series of tests for the commit logic 👍 . I also decided to change things slightly around handling committed values, so now you need to pass a value into `createCommit` and `confirmCommit` to (1) make sure that the editor value matches the value you're trying to commit and (2) make sure that the value you're trying to confirm matches the value you originally committed. And I actually decided to pull the synthetic diffing into this project after all, as I think it should be exclusively used when you receive remote commits. Moving it in here confines it to that single use case, which I think is good because we don't want to rely on it elsewhere (and it also makes testing easier). Finally I also exposed a `sendableCommits` method to show which committed values the editor is currently holding (also nice for testing). (Oh and I added the lock file to this repo, so it's inflating the diff size 😬)

Thanks for the reviews @annie + @nebrius!! Setting up a time to chat about this sounds great! will follow up with a calendar event so we can talk through things in more detail 👌

File: src/collab.ts [hunk 3]
```diff
@@ -108,64 +188,294 @@ export function receiveUpdates(
...
+
+  // If no commits are being confirmed, return old collab state.
+  if (!count) return state.update({annotations: [collabReceive.of(collabState)]})
+
+  // Determine which and how many updates are being confirmed.
+  const confirmedCommits = collabState.unconfirmedCommits.slice(0, count)
+
+  // Begin constructing new collab state.
+  let version = collabState.version
+  const unconfirmedUpdates = collabState.unconfirmedUpdates.slice()
```
Comment (nebrius) [inline src/collab.ts:389]:
```suggestion
  const unconfirmedUpdates = [...collabState.unconfirmedUpdates];
```

I assume we're doing this cause we want to clone the array, correct?

File: test/test-collab.ts [hunk 4]
```diff
@@ -1,37 +1,49 @@
...
 import {history, undo, redo, isolateHistory} from "@codemirror/history"
 import ist from "ist"
-import {collab, receiveUpdates, sendableUpdates, Update, getClientID, getSyncedVersion} from "@codemirror/collab"
-
+import {collab, receiveUpdates, createCommit, confirmCommits, receiveCommits, sendableCommits, sendableUpdates, Update, getClientID} from "@observablehq/collab"
 class DummyServer {
   states: EditorState[] = []
-  updates: Update[] = []
-  clientIDs: string[] = []
+  received: (string | Update)[][] = []
```
Comment (nebrius) [inline test/test-collab.ts:7]: I _think_ I get what's going with this array of arrays after reading the code, but it's not very clear from the TypeScript types alone. Could we drop in a quick comment explaining the purpose of the array nesting?

**Review by nebrius — APPROVED** [review]
Couple small nits, but this looks good!

For posterity from our call this morning, we're not going to do much in-depth review on this PR due to conceptual complexity plus our eventual goal of refactoring in a non-backwards-compatible-with-classic way once GA is out the door, and instead focus on testing the corresponding observablehq repo PR that links to this updated version.
